src/foam/nanos/fs/FileArrayDAODecorator.js

```diff
@@ -25,42 +25,51 @@ foam.CLASS({
       class: 'Long',
       name: 'maxStringDataSize',
       value: 1024 * 3
+    },
+    {
+      class: 'Boolean',
+      name: 'skipToData',
+      value: false
     }
   ],
 
   methods: [
-    function write(X, dao, obj, existing) {
-      var self  = this;
-      var props = obj.cls_.getAxiomsByClass(foam.nanos.fs.FileArray);
-
-      var promises = props.map((prop) => {
-        let files = prop.f(obj);
-        return Promise.all(files.map(async f => {
-
-          // We do not allow file update, so there is no point to send file again
-          // if it is already stored and has id
-          if ( f.id ) return f;
+    async function write(X, dao, obj, existing) {
+      var newObj = this.skipToData ? obj.data : obj;
+      if ( ! newObj ) {
+        return Promise.resolve(obj);
+      }
+      await this.processFiles(newObj);
+      return Promise.resolve(obj);
+    },
 
-          if ( f.filesize <= this.maxStringDataSize ) {
-            f.dataString = await this.encode(f.data.blob);
-            f.data = undefined;
-          } else {
-            f.dataString = undefined;
-          }
-          return self.fileDAO.put(f);
-        }));
-      });
+    async function processFiles(obj) {
+      var props1 = obj.cls_.getAxiomsByClass(foam.nanos.fs.FileArray);
+      var values = await Promise.all(props1.map(prop => Promise.all(prop.f(obj).map(f => this.processFile(f)))));
+      values.forEach(f => f.forEach(f2 => {
+        f2.dataString = undefined;
+        f2.data = undefined;
+      }));
+      props1.forEach((prop, i) => prop.set(obj, values[i]));
+      var props2 = obj.cls_.getAxiomsByClass(foam.core.FObjectProperty).filter((p) => ! foam.dao.DAOProperty.isInstance(p));
```

Comment:
Question: is a FileArray also a FObjectProperty? Curious if obj FileArray properties are processed twice?

Reply:
Yes, currently a nested FileArray would be processed twice... We can add another condition to the filter of props2 to be ! props1.includes(p).

Reply:
actually, I don't think my suggestion makes sense...

Reply:
actually, it does.